bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowerMaxBinding.java

```diff
@@ -0,0 +1,907 @@
+                try {
+                    panelType = PowerMaxPanelType.fromLabel((String) config.get("panelType"));
+                } catch (IllegalArgumentException exception) {
+                    panelType = PowerMaxPanelType.POWERMAX_PRO;
+                    logger.warn("PowerMax alarm binding: panel type not configured correctly");
+                }
+            }
+            PowerMaxPanelSettings.initPanelSettings(panelType);
+
+            if (StringUtils.isNotBlank((String) config.get("autoSyncTime"))) {
+                autoSyncTime = Boolean.valueOf((String) config.get("autoSyncTime"));
+            }
+
+            if (StringUtils.isNotBlank((String) config.get("pinCode"))) {
+                pinCode = (String) config.get("pinCode");
+            }
+
+            properlyConfigured = true;
+
+        } catch (ConfigurationException e) {
+
+            logger.warn("PowerMax alarm binding: {}", e.getReason());
+
+        } finally {
+
+            this.setProperlyConfigured(properlyConfigured);
+
+        }
+    }
+
+    /**
+     * Open a TCP or Serial connection to the PowerMax Alarm Panel
+     */
+    private void openConnection() {
+        PowerMaxCommDriver.initTheCommDriver(serialPort, ipAddress, tcpPort);
+        PowerMaxCommDriver comm = PowerMaxCommDriver.getTheCommDriver();
+        if (comm != null) {
+            comm.addEventListener(this);
+            connected = comm.open();
+            if (serialPort != null) {
+                logger.info("PowerMax alarm binding: serial connection ({}): {}", serialPort,
+                        connected ? "connected" : "disconnected");
+            } else if (ipAddress != null) {
+                logger.info("PowerMax alarm binding: TCP connection (IP {} port {}): {}", ipAddress, tcpPort,
+                        connected ? "connected" : "disconnected");
+            }
+        } else {
+            connected = false;
+        }
+        logger.debug("openConnection(): {}", connected ? "connected" : "disconnected");
+    }
+
+    /**
+     * Close TCP or Serial connection to the PowerMax Alarm Panel and remove the Event Listener
+     */
+    private void closeConnection() {
+        PowerMaxCommDriver comm = PowerMaxCommDriver.getTheCommDriver();
+        if (comm != null) {
+            comm.close();
+            comm.removeEventListener(this);
+        }
+        connected = false;
+        logger.debug("closeConnection(): disconnected");
+    }
+
+    /**
+     * PowerMax Alarm incoming message event handler
+     *
+     * @param event
+     */
+    @Override
+    public void powerMaxEventReceived(EventObject event) {
+        PowerMaxEvent powerMaxEvent = (PowerMaxEvent) event;
+        PowerMaxBaseMessage message = powerMaxEvent.getPowerMaxMessage();
+
+        logger.debug("powerMaxEventReceived(): received message {}", (message.getReceiveType() != null)
+                ? message.getReceiveType().getDescription() : String.format("%02X", message.getCode()));
+
```

Comment:
perhaps wrap above in `if (logger.isDebugEnabled()) { ... }` to avoid the `String.format` overhead on each call of this method under normal operation?